packages/flutter/lib/src/cupertino/refresh.dart

```diff
@@ -0,0 +1,520 @@
+                  size: 36.0,
+                ),
+              )
+            : new Opacity(
+                opacity: opacityCurve.transform(
+                  min(pulledExtent / refreshIndicatorExtent, 1.0)
+                ),
+                child: const CupertinoActivityIndicator(radius: 14.0),
+              ),
+      ),
+    );
+  }
+
+  @override
+  _CupertinoRefreshControlState createState() {
+    return new _CupertinoRefreshControlState();
+  }
+}
+
+class _CupertinoRefreshControlState extends State<CupertinoRefreshControl> {
+  /// Reset the state from done to inactive when only this fraction of the
+  /// original `refreshTriggerPullDistance` is left.
+  static const double _kInactiveResetOverscrollFraction = 0.1;
+
+  RefreshIndicatorMode refreshState;
+  // [Future] returned by the widget's `onRefresh`.
+  Future<void> refreshTask;
+  // The amount of space available from the inner indicator box's perspective.
+  //
+  // The value is the sum of the sliver's layout extent and the overscroll
+  // (which partially gets transfered into the layout extent when the refresh
+  // triggers).
+  //
+  // The value is independent from the slivers scrollOffset. i.e. it's still
+  // the same when the sliver scrolls away without retracting.
+  double lastIndicatorExtent = 0.0;
+  bool hasSliverLayoutExtent = false;
+
+  @override
+  void initState() {
+    super.initState();
+    refreshState = RefreshIndicatorMode.inactive;
+  }
+
+  // A state machine transition calculator. Multiple states can be transitioned
+  // through per single call.
+  RefreshIndicatorMode transitionNextState() {
+    RefreshIndicatorMode nextState;
+
+    void goToDone() {
+      nextState = RefreshIndicatorMode.done;
+      // Either schedule the RenderSliver to re-layout on the next frame
+      // when not currently in a frame or schedule it on the next frame.
+      if (SchedulerBinding.instance.schedulerPhase == SchedulerPhase.idle) {
+        setState(() => hasSliverLayoutExtent = false);
+      } else {
+        SchedulerBinding.instance.addPostFrameCallback((Duration timestamp){
+          setState(() => hasSliverLayoutExtent = false);
+        });
+      }
+    }
+
+    switch (refreshState) {
+      case RefreshIndicatorMode.inactive:
+        if (lastIndicatorExtent <= 0) {
+          return RefreshIndicatorMode.inactive;
+        } else {
+          nextState = RefreshIndicatorMode.drag;
+        }
+        continue drag;
+      drag:
+      case RefreshIndicatorMode.drag:
+        if (lastIndicatorExtent == 0) {
+          return RefreshIndicatorMode.inactive;
+        } else if (lastIndicatorExtent < widget.refreshTriggerPullDistance) {
+          return RefreshIndicatorMode.drag;
+        } else {
+          if (widget.onRefresh != null) {
+            HapticFeedback.mediumImpact();
+            // Call onRefresh after this frame finished since the function is
+            // user supplied and we're always here in the middle of a the sliver's
```

Comment:
of a the --> of the
(or 'of a', depending on what you meant)

Reply:
ah thanks